presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java

```diff
@@ -0,0 +1,147 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spiller;
+
+import com.facebook.presto.operator.exchange.LocalPartitionGenerator;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import io.airlift.concurrent.MoreFutures;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class GenericPartitioningSpiller
+        implements PartitioningSpiller
+{
+    private final List<Type> types;
+    private final int partitionsCount;
+    private final boolean[] ignorePartitions;
+    private final PageBuilder[] pageBuilders;
+    private final LocalPartitionGenerator partitionGenerator;
+    private final SingleStreamSpiller[] spillers;
+    private boolean readingStarted;
+
+    public GenericPartitioningSpiller(
+            List<Type> types,
+            LocalPartitionGenerator partitionGenerator,
+            int partitionsCount,
+            Set<Integer> ignorePartitions,
+            SpillerFactory spillerFactory)
+    {
+        this.partitionsCount = partitionsCount;
+        this.types = requireNonNull(types, "types is null");
+        this.spillers = new SingleStreamSpiller[partitionsCount];
+        this.ignorePartitions = new boolean[partitionsCount];
+        this.pageBuilders = new PageBuilder[partitionsCount];
+        this.partitionGenerator = requireNonNull(partitionGenerator, "partitionGenerator is null");
+
+        for (int partition = 0; partition < partitionsCount; partition++) {
+            this.ignorePartitions[partition] = ignorePartitions.contains(partition);
+            pageBuilders[partition] = new PageBuilder(types);
+            spillers[partition] = spillerFactory.createSingleStreamSpiller(types);
+        }
+    }
+
+    @Override
+    public synchronized Iterator<Page> getSpilledPages(int partition)
+    {
+        readingStarted = true;
+        checkArgument(partition < partitionsCount);
+        getFutureValue(flush(partition));
+        return spillers[partition].getSpilledPages();
+    }
+
+    @Override
+    public synchronized PartitioningSpillResult partitionAndSpill(Page page)
+    {
+        checkState(!readingStarted);
+        IntArrayList unspilledPositions = partitionPage(page);
+        CompletableFuture<?> future = flush();
```

Comment:
it already works as you want. Pages are only flushed when their PageBuilder `isFull`